src/Ledger/Foreign/HSLedger.agda

```diff
@@ -72,23 +72,22 @@ module Implementation where
   open import Ledger.TokenAlgebra ℕ
   coinTokenAlgebra : TokenAlgebra
   coinTokenAlgebra = λ where
-    .Value-CommutativeMonoid   → +-0-commutativeMonoid
-    .coin                      → id
-    .inject                    → id
-    .policies                  → λ _ → ∅
-    .size                      → λ x → 1 -- there is only ada in this token algebra
-    ._≤ᵗ_                      → _≤_
-    .AssetName                 → String
-    .specialAsset              → "Ada"
-    .property                  → λ _ → refl
-    .coinIsMonoidHomomorphism → λ where
-      .isMagmaHomomorphism → λ where
-        .isRelHomomorphism → record {cong = id}
-        .homo → λ _ _ → refl
-      .ε-homo → refl
+    .Value                      → ℕ
+    .Value-IsCommutativeMonoid' → it
+      -- ^ Agda bug? Without this line, `coinIsMonoidHomomorphism` doesn't type check anymore
```

Comment:
Here's something really stupid. Typeclass resolution will find this if the line is removed, and I checked, `Value-IsCommutativeMonoid'` normalizes to the same expression (`toCommMonoid' {zeroˡ} {zeroˡ} +-0-commutativeMonoid`) in either case.

But `coinIsMonoidHomomorphism` won't work if the line is removed. The reason is that when we provide the morphism explicitly like in the old version, this line breaks:
```agda
.isRelHomomorphism → record {cong = id}
```
What's the type of `cong`? If you replace `id` with `λ {x} {y} h → ?` and normalize the type, it'll be `h : x ≡ y ⊢ x ≡ y`, and Agda will accept `h` interactively, but when you re-check it it'll claim that `h` doesn't have the type `x ≡ y`. I also tried stuff like `let h' = x ≡ y ∋ h in h` to no avail.

@omelkonian @UlfNorell any ideas?

Reply:
My suspicion is that if this line is left out, Agda instantiates the typeclass later than with this line, which leads Agda to give up too early in trying to compare the two equality types.

Reply:
That's strange, probably your suspicion is correct that it's an issue with the timing of instance resolution (as most often is). 

A quick fix for now is to simply replace the back-and-forth `toCommMonoid'/fromCommMonoid'` when opening `CommutativeMonoid` to get `rawMonoid` with an immediate `open CommutativeMonoid ℕ.+-0-commutativeMonoid using (rawMonoid)` and the problem goes away.

I will continue investigating to report a minimal Agda bug.

Reply:
Reported as [Agda issue #7038](https://github.com/agda/agda/issues/7038); in this minimal example replacing co-patterns with `record` notation solves the error, but not in our case!